periodic/periodic.go

```diff
@@ -219,9 +233,24 @@ func runOne(id int, funcTimes *stats.Histogram, sleepTimes *stats.Histogram, num
 	perThreadQPS := r.QPS / float64(r.NumThreads)
 	useQPS := (perThreadQPS > 0)
 	f := r.Function
+
+	// Catch SIGINT signals from the OS and raise a flag to terminate the run loop
+	terminated := false
+	c := make(chan os.Signal, 1)
+	signal.Notify(c, os.Interrupt)
+	go func() {
+		<-c
+		terminated = true
+	}()
+
 	for {
+		if terminated {
```

Comment:
can you run fortio -qps 0 against itself on your machine before/after this change, I would like to know how expensive the channel/synchronization is

Reply:
Although sync is no longer used due to use of switch on the channel instead of a go routine here is the load output for 5s run with -qps 0 **after** the last commit:

Before:

> Fortio v0.2.8 running at 0 queries per second, 8->8 procs, for 5s: http://localhost:8080
> Starting at max qps with 4 thread(s) [gomax 8] for 5s
> 13:59:23 I httprunner.go:79> Starting http test for http://localhost:8080 with 4 threads at 0.0 qps
> Ended after 5.000131225s : 206945 calls. qps=41388
> Aggregated Function Time : count 206945 avg 9.6066258e-05 +/- 2.04e-05 min 4.4506e-05 max 0.000975281 sum 19.8804318
> \# range, mid point, percentile, count
> < 0.001 , 0.001 , 100.00, 206945
> \# target 50% 0.000509894
> \# target 75% 0.000742587
> \# target 99% 0.000965973
> \# target 99.9% 0.00097435
> Code 200 : 206945
> Response Header Sizes : count 206945 avg 116 +/- 0 min 116 max 116 sum 24005620
> Response Body/Total Sizes : count 206945 avg 116 +/- 0 min 116 max 116 sum 24005620
> All done 206945 calls (plus 4 warmup) 0.096 ms avg, 41387.9 qps
> 13:59:28 I periodic.go:257> T003 ended after 5.000038522s : 51711 calls. qps=10342.120320168206
> 13:59:28 I periodic.go:257> T001 ended after 5.000097966s : 51763 calls. qps=10352.397163411897
> 13:59:28 I periodic.go:257> T002 ended after 5.000099538s : 51729 calls. qps=10345.59404405201
> 13:59:28 I periodic.go:257> T000 ended after 5.000101263s : 51742 calls. qps=10348.190422238655
> 

After:

> Fortio v0.3.0 running at 0 queries per second, 8->8 procs, for 5s: http://localhost:8080
> 13:58:10 I httprunner.go:79> Starting http test for http://localhost:8080 with 4 threads at 0.0 qps
> Starting at max qps with 4 thread(s) [gomax 8] for 5s
> Ended after 5.000094758s : 207429 calls. qps=41485
> Aggregated Function Time : count 207429 avg 9.5778898e-05 +/- 1.952e-05 min 4.632e-05 max 0.000439454 sum 19.867321
> \# range, mid point, percentile, count
> < 0.001 , 0.001 , 100.00, 207429
> \# target 50% 0.000242887
> \# target 75% 0.000341171
> \# target 99% 0.000435523
> \# target 99.9% 0.000439061
> Code 200 : 207429
> Response Header Sizes : count 207429 avg 116 +/- 0 min 116 max 116 sum 24061764
> Response Body/Total Sizes : count 207429 avg 116 +/- 0 min 116 max 116 sum 24061764
> All done 207429 calls (plus 4 warmup) 0.096 ms avg, 41485.0 qps
> 13:58:15 I periodic.go:296> T000 ended after 5.000007637s : 51894 calls. qps=10378.784147445092
> 13:58:15 I periodic.go:296> T002 ended after 5.000052854s : 51799 calls. qps=10359.69048978377
> 13:58:15 I periodic.go:296> T003 ended after 5.000057289s : 51867 calls. qps=10373.281145019297
> 13:58:15 I periodic.go:296> T001 ended after 5.000056364s : 51869 calls. qps=10373.683059545607
>